url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/environment/geometry/InfinitePlane.java

```diff
@@ -0,0 +1,300 @@
+    {
+        this.clone(other);
+    }
+
+    public InfinitePlane(double a, double b, double c, double d) {
+        this.a = a;
+        this.b = b;
+        this.c = c;
+        this.d = d;
+    }
+
+    public InfinitePlane(Vector3D normal, Vector3D pointInPlane) {
+        normal.normalize();
+        a = normal.x;
+        b = normal.y;
+        c = normal.z;
+        d = -normal.dotProduct(pointInPlane);
+    }
+
+    public InfinitePlane(Vector3D p0, Vector3D p1, Vector3D p2) {
+        Vector3D a, b, normal;
+        a = p1.substract(p0);
+        a.normalize();
+        b = p2.substract(p0);
+        b.normalize();
+        normal = a.crossProduct(b);
+        normal.normalize();
+        this.a = normal.x;
+        this.b = normal.y;
+        this.c = normal.z;
+        this.d = -normal.dotProduct(p0);
+    }
+
+    public void clone(InfinitePlane other)
+    {
+        this.a = other.a;
+        this.b = other.b;
+        this.c = other.c;
+        this.d = other.d;
+    }
+
+    public boolean
+    doIntersection(Ray inout_rayo) {
+        double denominator = a*inout_rayo.direction.x + b*inout_rayo.direction.y + c*inout_rayo.direction.z;
+        if ( Math.abs(denominator) < VSDK.EPSILON ) return false;
+        double t = -(a*inout_rayo.origin.x + b*inout_rayo.origin.y + c*inout_rayo.origin.z + d)/denominator;
+
+        if ( t < 0 ) return false;
+
+        inout_rayo.t = t;
+
+        return true;
+    }
+
+    public boolean
+    doIntersectionWithNegative(Ray inout_rayo) {
+        double denominator = a*inout_rayo.direction.x + b*inout_rayo.direction.y + c*inout_rayo.direction.z;
+        if ( Math.abs(denominator) < VSDK.EPSILON ) {
+            Ray r = new Ray(inout_rayo.origin, inout_rayo.direction.multiply(-1));
+            if ( doIntersection(r) ) {
+                inout_rayo.t = -r.t;
+                return true;
+            }
+            else {
+                return false;
+            }
+        }
+        double t = -(a*inout_rayo.origin.x + b*inout_rayo.origin.y + c*inout_rayo.origin.z + d)/denominator;
+
+        inout_rayo.t = t;
+
+        return true;
+    }
+
+    /**
+    Por a given point `p`, calculates if it lies inside, outside or 
+    on surface with respect to current plane, taking the plane as an
+    infinite halfspace, not as a surface.
+    @return 0 if point is on the plane surface, 1 if point is outside or
+    -1 if point is inside the plane.
+    Note that current interpretation of the plane is done as a semispace,
+    where "outside" means the direction pointed by plane's normal.
+    */
+    public int doContainmentTestHalfSpace(Vector3D p,
+                                          double distanceTolerance) {
+        double num = a*p.x + b*p.y + c*p.z + d;
+        int op = LIMIT;
+
+        if( num > distanceTolerance ) {
+            op = OUTSIDE;
+        }
+        else if( num < -distanceTolerance ) {
+            op = INSIDE;
+        }
+        return op;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 111:120, 129:138.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.